### PR TITLE
implement an asynchronous, data driven architecture

### PR DESCRIPTION
**Pull request by nrdxp**
## Impetus
This should drastically reduce the complexity of the implementation by describing all the relevant information we need into a single unified `BitteCluster` struct.

In addition, the current implementation relies on synchronous, blocking API calls, which is why many of the commands are painfully slow. By making proper use of the multi-threaded tokio runtime and the asynchronous reqwest library, we can drastically reduce this burden by making each call in parallel.

Also, the cli will now start building this cluster defining struct at the very beginning of the main function in a separate thread, in order to keep the waiting to a bare minimum.

As an example of why it is awesome and useful to have our own data format, checkout how I obtain the alloc index even though the Nomad API doesn't usually expose it directly:
https://github.com/input-output-hk/bitte-cli/blob/f0d0647bbf3d8df5f108c0c488e3da568481f065/lib/src/types.rs#L612-L627

## Tasks:
- [x] Initial defintion of a `BitteCluster` struct (subject to feedback)
- [x] Make all API calls asynchronously and in parallel
  - [x] Implement async function to pull the relevant Nomad allocation information.
  - [x] Implement async function to pull the relevant Nomad node information.
  - [x] Implement async function to pull the relevant bitte cluster node information, e.g. IP, etc.
- [x] Refactor the entire codebase to rely on this `BitteCluster` struct instead of making ad hoc API calls everywhere
- [x] Serialize the `BitteCluster` struct to a known location so Nix can import it and automatically define deploy-rs nodes. see #18
- [x] Implement a semantic versioning scheme once this PR is ready, so we can have a sane development process :smile:
- [x] fix darwin build

## Stretch goals:
- [x] Use the serialized definition as a cache to avoid unecesary API calls when running the cli in quick succession
- [x] Improve cli semantics by allow for things like `bitte ssh $job $task $index`
- [ ] Use the serialization to generate cluster aware cli completions

> ### Note:
> The `BitteProvider` enum only has a single value atm, but it's there to make things simpler when we move everything to on-premise, while still maintaining compatibility with AWS. This also leaves open the option to add other cloud platforms in the future by simply implementing the `find_bitte_nodes()` function for that provider.

### CODE REVIEW

File: lib/src/types.rs [hunk 1]
```diff
@@ -545,3 +558,125 @@ pub struct TerraformStateRoles {
...
+                    .default_headers(headers)
+                    .gzip(true)
+                    .build()?,
+            )
+        };
+        let allocs = tokio::spawn(BitteCluster::find_allocs(
+            Arc::clone(&nomad_client),
+            domain.clone(),
+        ));
+        let nodes = tokio::spawn(BitteCluster::find_nodes(provider.clone()));
```
Comment (jonringer) [inline lib/src/types.rs:652]:
```suggestion
        let nodes = tokio::spawn(BitteCluster::find_nodes(provider);
```

if you add `#[derive(Copy)]` to provider, then the clone should be unnecessary.

File: lib/src/types.rs [hunk 2]
```diff
@@ -545,3 +558,125 @@ pub struct TerraformStateRoles {
...
+    pub name: String,
+    pub nodes: Vec<BitteNode>,
+    pub allocs: NomadAllocs,
+    pub domain: String,
+    pub provider: BitteProvider,
+    #[serde(skip)]
+    pub nomad_client: Arc<Client>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
```
Comment (jonringer) [inline lib/src/types.rs:574]:
```suggestion
#[derive(Debug, Serialize, Deserialize, Copy, Clone)]
```

File: lib/src/types.rs (hunk 1 above)
Comment (nrdxp) [reply lib/src/types.rs]: fixed it :+1:

File: lib/src/types.rs [hunk 3]
```diff
@@ -545,3 +570,436 @@ pub struct TerraformStateRoles {
...
+impl BitteCluster {
+    pub async fn new() -> anyhow::Result<Self> {
+        let name = env::var("BITTE_CLUSTER")?;
+        let domain = env::var("BITTE_DOMAIN")?;
+        let provider: BitteProvider = {
+            let string = env::var("BITTE_PROVIDER")?;
+            match string.parse() {
+                Ok(v) => Ok(v),
+                Err(_) => Err(Error::ProviderError { provider: string }),
+            }?
```
Comment (jonringer) [inline lib/src/types.rs:903]: Touching of ENV, and validation should probably occur outside of a constructor call, generally you just want to create internally consistent logic, not have a bunch of business logic :)

Then again, if I were to re-write this, It would probably be named just slightly different and do the same thing, so I'm not sure.

**Review by jonringer — APPROVED** [review]

File: cli/src/cli.rs [hunk 4]
```diff
@@ -45,13 +47,85 @@ pub(crate) async fn provision(sub: &ArgMatches) -> Result<()> {
...
+pub(crate) async fn ssh(sub: &ArgMatches, cluster: ClusterHandle) -> Result<()> {
     let mut args = sub.values_of_lossy("args").unwrap_or_default();
+    let job: Vec<String> = sub.values_of_t("job").unwrap_or_default();
 
-    let ip = find_instance(needle.as_str())
-        .await
-        .map_or_else(|| needle.clone(), |i| i.public_ip);
+    let namespace: String = sub
+        .value_of_t("namespace")
+        .unwrap_or(env::var("NOMAD_NAMESPACE")?);
```
Comment (manveru) [inline cli/src/cli.rs:56]: I'd replace most `unwrap`s in this PR with `expect` so we have some idea of wtf goes on when it fails...

File: lib/src/rebuild.rs [hunk 5]
```diff
@@ -25,35 +49,37 @@ pub async fn copy(only: Vec<&str>, delay: Duration, copy: bool) -> Result<()> {
...
+        flake, instance.nixos
     );
-    let rebuild_flake: String = format!("{}#{}", flake, instance.uid);
+    let rebuild_flake: String = format!("{}#{}", flake, instance.nixos);
 
     nix_build(&target)?;
-    if copy {
+
+    if let Some(c) = cache {
+        let cache = format!("{}&secret-key=secrets/nix-secret-key-file", c);
```
Comment (manveru) [inline lib/src/rebuild.rs:74]: I'd love to fetch this value from some file in the encrypted folder instead so we can slowly get rid of the secrets dir.

File: package.nix [hunk 6]
```diff
@@ -1,23 +1,23 @@
...
 , zlib
+, rustPlatform
 
-# darwin dependencies
+  # darwin dependencies
 , darwin
 }:
 
-naersk.buildPackage {
+rustPlatform.buildRustPackage {
```
Comment (manveru) [inline package.nix:12]: What's the reason for getting rid of naersk? I'm not a big fan of `buildRustPackage` since it's very annoying to keep the hash updated and takes a long time to calculate it.

File: flake.nix [hunk 7]
```diff
@@ -31,26 +26,28 @@
...
-            rustfmt
-          ] ++ lib.optionals stdenv.isDarwin (with darwin; with apple_sdk.frameworks; [
-            libiconv
-            libresolv
-            Libsystem
-            SystemConfiguration
-            Security
-            CoreFoundation
-          ]);
+          buildInputs = with pkgs;
```
Comment (manveru) [inline flake.nix:29]: nixfmt pls

File: flake-compat.nix [hunk 8]
```diff
@@ -1,6 +1,7 @@
 # The `default.nix` in flake-compat reads `flake.nix` and `flake.lock` from `src` and
 # returns an attribute set of the shape `{ defaultNix, shellNix }`
 
-import (fetchTarball https://github.com/edolstra/flake-compat/archive/master.tar.gz) {
-  src = ./.;
-}
+import (fetchTarball
```
Comment (manveru) [inline flake-compat.nix:4]: why this change? also do we need that file at all?

File: cli/src/cli.rs [hunk 9]
```diff
@@ -218,73 +298,39 @@ pub async fn terraform_apply(workspace: String, _sub: &ArgMatches) -> Result<()>
...
+            .map(|mut node| {
+                node.nomad_client = None;
+                node
+            })
+            .collect();
+        cluster.nodes = nodes;
+        serde_json::to_writer_pretty(handle, &cluster)?;
+    } else {
+        let mut instance_table = Table::new();
+        instance_table.add_row(row!["Name", "Private IP", "Public IP"]);
```
Comment (manveru) [inline cli/src/cli.rs:318]: Not a fan of getting rid of all this information... i'd like at least the type, region, and suffix to remain. the rest could be in the json output.

**Review by manveru — APPROVED** [review]
Nothing blocking the merge since I think the benefits outweigh the issues I found, but would be good if we could fix them rather soon anyway.

File: package.nix (hunk 6 above)
Comment (nrdxp) [reply package.nix:12]: Yeah I did this reluctantly because naersk doesn't support downloading git dependencies in pure mode. The strange thing is that this didn't seem to affect hydra so I didn't notice it right away. But if anyone needed to build the nix package locally it would fail. I did [request](https://github.com/serokell/deploy-rs/pull/104#issuecomment-892803736) that they publish deploy-rs on crates.io since they merged my changes so we'll see.

I also wanted to give crates2nix a go and see if perhaps that would work.

File: flake-compat.nix (hunk 8 above)
Comment (nrdxp) [reply flake-compat.nix:4]: I think it was just an auto-format.This would be for stable nix users, if they want to build the package. Might be worth supporting that use case until flakes become the default.

File: cli/src/cli.rs (hunk 9 above)
Comment (nrdxp) [reply cli/src/cli.rs:318]: I had a feeling someone would complain, but I also felt like the previous table was a bit overwhelming, so my strategy was to remove almost everything and wait for someone to complain :sweat_smile: 

What you request sounds reasonable.

File: cli/src/cli.rs (hunk 4 above)
Comment (nrdxp) [reply cli/src/cli.rs:56]: Absolutely agree. There are a lot of things I wanted to clean up here, and this was one of them. I'd also like to make the imports a bit more idiomatic, and see if I could optimize some of the find_needle functions, but I think it'd be better to merge and then add this later. But the error messages are important enough, that I can probably get to that over the weekend sometime :crossed_fingers:

File: lib/src/types.rs (hunk 3 above)
Comment (nrdxp) [reply lib/src/types.rs:903]: Fair point. My first implementation actually did pass the environment in, but it added a function simply to call the constructor, which I felt was kinda overkill, and I decided to add it in, because the struct cannot be built without this information and so seems like it's not too horrible here.

File: package.nix (hunk 6 above)
Comment (manveru) [reply package.nix:12]: Ah yeah, the long-standing issue of recursive git dependencies has been a constant pain in our other rust repos too... the easy way is to simply add the dependency to our own and naersk will resolve it just fine.

File: flake-compat.nix (hunk 8 above)
Comment (manveru) [reply flake-compat.nix:4]: I somewhat doubt any of this cli would work without flakes... but ok.

Comment (nrdxp) [reply flake-compat.nix:4]: Actually yeah, now that you mention that, it probably doesn't make sense :sweat_smile:

File: cli/src/cli.rs (hunk 9 above)
Comment (nrdxp) [reply cli/src/cli.rs:304]: After some additional reflection, some of the requested fields may not be available when on prem. I suppose I could still implement this and just construct the table fields based on the BitteProvider

File: cli/src/cli.rs (hunk 4 above)
Comment (nrdxp) [reply cli/src/cli.rs:56]: @manveru, how would you feel about simply checking the `IN_NIX_SHELL` var (and maybe some other BITTE_* var as well) at the beginning of main, and provide an error message if it doesn't exist advising that one must run bitte-cli from a bitte shell?

File: package.nix (hunk 6 above)
Comment (nrdxp) [reply package.nix:6]: looks like nix-community/naersk#167 actually resolves the issue. I'll force push an update that removes my old fix and bumps naersk to this PR instead.
